aiosmtpd/tests/test_starttls.py

```diff
@@ -0,0 +1,219 @@
+import asyncio
+import os
+import unittest
+from email.mime.text import MIMEText
+from smtplib import SMTP
+try:
+    import ssl
+    from asyncio import sslproto
+except ImportError:
+    _has_ssl = False
+else:
+    _has_ssl = True
+
+from .test_smtp import Controller as OldController, Sink
+
+from aiosmtpd.smtp import SMTP as SMTPProtocol
+
+
+class Controller(OldController):
+    def factory(self):
+        return SMTPProtocol(self.handler)
+
+
+class ReceivingHandler:
+    box = None
+
+    @asyncio.coroutine
+    def process_message(self, *args, **kws):
+        if not self.box:
+            self.box = []
+        self.box.append(args)
+
+
+def get_tls_context():
+    tls_context = ssl.create_default_context(ssl.Purpose.CLIENT_AUTH)
+    path = os.path.dirname(__file__)
+    path = os.path.join(path, 'certs')
+    tls_context.load_cert_chain(
+        os.path.join(path, 'server.crt'),
+        os.path.join(path, 'server.key')
+    )
+    return tls_context
+
+
+class TLSRequiredController(Controller):
+    def factory(self):
+        return SMTPProtocol(
+            self.handler,
+            decode_data=True,
+            require_starttls=True,
+            tls_context=get_tls_context()
+        )
+
+
+class TLSController(Controller):
+    def factory(self):
+        return SMTPProtocol(
+            self.handler,
+            decode_data=True,
+            require_starttls=False,
+            tls_context=get_tls_context()
+        )
+
+
+class HandshakeFailingHandler:
+    @staticmethod
+    def handle_tls_handshake(*args):
+        return False
+
+
+class TestStartTLS(unittest.TestCase):
+    if _has_ssl:
+        def test_starttls(self):
```

Comment:
I think it would be better to just skip these tests if `_has_ssl` is false.  Better to report them as skipped than to just not execute them at all.  E.g. use `unittest.skipIf(not _has_ssl, 'SSL is not available')`